sdk/program/src/bpf_loader_upgradeable.rs

```diff
@@ -0,0 +1,185 @@
+//! to update the program at any time.  This ability break's the "code is law"
+//! contract the usually enforces the policy that once a program is on-chain it
+//! becomes immutable.  Because of this, care should be taken before executing
+//! upgradeable programs which still have a functioning authority.  For more
+//! information refer to `loader_upgradeable_instruction.rs`
+
+use crate::{
+    instruction::{AccountMeta, Instruction, InstructionError},
+    loader_upgradeable_instruction::UpgradeableLoaderInstruction,
+    pubkey::Pubkey,
+    system_instruction, sysvar,
+};
+use bincode::serialized_size;
+
+crate::declare_id!("BPFLoaderUpgradeab1e11111111111111111111111");
+
+#[derive(Debug, Serialize, Deserialize, PartialEq, Clone, Copy, AbiExample)]
+pub enum UpgradableLoaderState {
+    /// Account is not initialized.
+    Uninitialized,
+    /// An executable account.
+    Executable {
+        /// Public key of the code account.
+        code_pubkey: Pubkey,
+    },
+    // A code account.
+    Code {
+        /// Slot that the program was last modified.
+        slot: u64,
+        /// Program's authority.
+        authority: Option<Pubkey>,
+        // The raw executable code follows this serialized structure in the
+        // account's data.
+    },
+}
+
+/// Size of an executable account's data.
+pub fn size_of_executable() -> Result<u64, InstructionError> {
+    serialized_size(&UpgradableLoaderState::Executable {
+        code_pubkey: Pubkey::default(),
+    })
+    .map_err(|_| InstructionError::InvalidInstructionData)
+}
+/// Size of a code account's data.
+pub fn size_of_code(program_size: u64) -> Result<u64, InstructionError> {
+    Ok(serialized_size(&UpgradableLoaderState::Code {
+        slot: 0,
+        authority: Some(Pubkey::default()),
+    })
+    .map_err(|_| InstructionError::InvalidInstructionData)?
+        + program_size)
+}
+/// Offset into the code account's data of the program bits.
+pub fn offset_of_program_data() -> Result<u64, InstructionError> {
+    size_of_code(0)
+}
+
+/// Returns the instructions required to create a source account
+pub fn create_source_account(
+    payer_pubkey: &Pubkey,
+    source_pubkey: &Pubkey,
+    lamports: u64,
+    program_size: u64,
+) -> Result<Vec<Instruction>, InstructionError> {
+    Ok(vec![system_instruction::create_account(
+        payer_pubkey,
+        source_pubkey,
+        lamports,
+        program_size,
+        &id(),
+    )])
+}
+
+/// Returns the instructions required to write a chunk of program data to a source account
+pub fn write(
+    source_pubkey: &Pubkey,
+    offset: u32,
+    bytes: Vec<u8>,
+) -> Result<Vec<Instruction>, InstructionError> {
+    Ok(vec![Instruction::new(
+        id(),
+        &UpgradeableLoaderInstruction::Write { offset, bytes },
+        vec![AccountMeta::new(*source_pubkey, false)],
+    )])
+}
+
+/// Returns the instructions required to deploy a program
+pub fn deploy(
```

Comment:
We hope that once realloc comes in `max_program_size` will no longer be required, so how about reserving `deploy()` for that sunny day and calling this function:
```suggestion
pub fn deploy_with_max_program_size(
```